lib/tls-gnutls.c

```diff
@@ -348,26 +673,100 @@ tls_handshake(tls_t *tls, int fd, bool client, const tls_handshake_t *hs)
       goto error;
   }
 
-  if (client && hs->clt.psk) {
-    ret = g2e(gnutls_psk_allocate_client_credentials(&tls->creds.clt.psk));
-    if (ret < 0)
-      goto error;
-
-    gnutls_psk_set_client_credentials_function(tls->creds.clt.psk, psk_clt);
-    ret = g2e(gnutls_credentials_set(tls->session, GNUTLS_CRD_PSK,
-                                     tls->creds.clt.psk));
-    if (ret < 0)
-      goto error;
-  } else if (!client && hs->srv.psk) {
-    ret = g2e(gnutls_psk_allocate_server_credentials(&tls->creds.srv.psk));
-    if (ret < 0)
-      goto error;
-
-    gnutls_psk_set_server_credentials_function(tls->creds.srv.psk, psk_srv);
-    ret = g2e(gnutls_credentials_set(tls->session, GNUTLS_CRD_PSK,
-                                     tls->creds.srv.psk));
-    if (ret < 0)
-      goto error;
+  if (client) {
```

Comment:
The control flow from here on down is hard to follow.  If there isn't a better way to write it, could you add some comments?